### PR TITLE
[ACE6] Fixes for IPv6 multicast and inet_ntop/inet_pton on Windows

### PR DESCRIPTION
**Pull request by kuznetsovmoci**
Changed logic for GetAdaptersAddresses as per Microsoft guidelines:
https://docs.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-getadaptersaddresses

Fixed definitions for
```
#define ACE_LACKS_INET_NTOP
#define ACE_LACKS_INET_PTON
```

### CODE REVIEW

File: ACE/ace/SOCK_Dgram.cpp [hunk 1]
```diff
@@ -786,21 +786,34 @@ ACE_SOCK_Dgram::make_multicast_ifaddr6 (ipv6_mreq *ret_mreq,
         ACE_OS::ace_isdigit (net_if[0]) &&
         (if_ix = ACE_OS::atoi (net_if)) > 0;
 
-      IP_ADAPTER_ADDRESSES tmp_addrs;
-      // Initial call to determine actual memory size needed
-      ULONG bufLen = 0;
-      char *buf = 0;
-      if (::GetAdaptersAddresses (AF_INET6, 0, 0, &tmp_addrs, &bufLen)
-          == ERROR_BUFFER_OVERFLOW)
+      ULONG bufLen = 15000; // Initial size as per Microsoft
```
Comment (jwillemsen) [inline ACE/ace/SOCK_Dgram.cpp:789]: The previous solution to query for the size of the buffer looks more optimal

Comment (jwillemsen) [reply ACE/ace/SOCK_Dgram.cpp:789]: What is the problem you have that require these changes?

Comment (kuznetsovmoci) [reply ACE/ace/SOCK_Dgram.cpp:789]: I am not sure. Microsoft says this call is expensive, and 15000 guesstimate is a good start.

Comment (kuznetsovmoci) [reply ACE/ace/SOCK_Dgram.cpp:789]: What's more important, is that the initial size obtained by the first call, in some cases is still insufficient for the next call. The example suggested in the article is what Microsoft recommends.

Comment (jwillemsen) [reply ACE/ace/SOCK_Dgram.cpp:789]: Ok, so it is a small performance optimization, not a real fix, code change is from https://docs.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-getadaptersaddresses?

Comment (kuznetsovmoci) [reply ACE/ace/SOCK_Dgram.cpp:789]: It is a real fix. `GetAdaptersAddresses` was failing on the second trial with `ERROR_BUFFER_OVERFLOW`.

Comment (jwillemsen) [reply ACE/ace/SOCK_Dgram.cpp:789]: Ok

Comment (kuznetsovmoci) [reply ACE/ace/SOCK_Dgram.cpp:789]: I was able to capture the exact behavior before and after the fix.

**Comment by jwillemsen** [comment]
Also open a PR with the same change for master

**Comment by kuznetsovmoci** [comment]
> Also open a PR with the same change for master

Sure. I am new to the process - should I have just opened this PR to the master, or should I have both PRs open in parallel?

**Comment by jwillemsen** [comment]
You can open both in paralllel when you want to have the change in ACE 7x and ACE 6.x

File: ACE/ace/SOCK_Dgram.cpp [hunk 2]
```diff
@@ -786,21 +786,34 @@ ACE_SOCK_Dgram::make_multicast_ifaddr6 (ipv6_mreq *ret_mreq,
...
+      char* buf = 0;
```
Comment (mitza-oci) [inline ACE/ace/SOCK_Dgram.cpp:790]:
```suggestion
      char *buf = 0;
```

File: ACE/ace/SOCK_Dgram.cpp [hunk 3]
```diff
@@ -786,21 +786,34 @@ ACE_SOCK_Dgram::make_multicast_ifaddr6 (ipv6_mreq *ret_mreq,
...
+      ACE_NEW_RETURN(buf, char[bufLen], -1);
```
Comment (mitza-oci) [inline ACE/ace/SOCK_Dgram.cpp:791]:
```suggestion
      ACE_NEW_RETURN (buf, char[bufLen], -1);
```

File: ACE/ace/SOCK_Dgram.cpp [hunk 4]
```diff
@@ -786,21 +786,34 @@ ACE_SOCK_Dgram::make_multicast_ifaddr6 (ipv6_mreq *ret_mreq,
...
+      DWORD dwRetVal = 0;
+      ULONG iterations = 0;
+      const ULONG maxTries = 3;
+      PIP_ADAPTER_ADDRESSES pAddrs;
+      do
         {
-          ACE_NEW_RETURN (buf, char[bufLen], -1);
-        }
+          pAddrs = reinterpret_cast<PIP_ADAPTER_ADDRESSES> (buf);
+          dwRetVal = ::GetAdaptersAddresses(AF_INET6, 0, 0, pAddrs, &bufLen);
```
Comment (mitza-oci) [inline ACE/ace/SOCK_Dgram.cpp:799]:
```suggestion
          dwRetVal = ::GetAdaptersAddresses (AF_INET6, 0, 0, pAddrs, &bufLen);
```

File: ACE/ace/SOCK_Dgram.cpp [hunk 5]
```diff
@@ -786,21 +786,34 @@ ACE_SOCK_Dgram::make_multicast_ifaddr6 (ipv6_mreq *ret_mreq,
...
+          if (dwRetVal == ERROR_BUFFER_OVERFLOW)
+            {
+              delete[] buf;
+              ACE_NEW_RETURN(buf, char[bufLen], -1);
```
Comment (mitza-oci) [inline ACE/ace/SOCK_Dgram.cpp:803]:
```suggestion
              ACE_NEW_RETURN (buf, char[bufLen], -1);
```

File: ACE/ace/SOCK_Dgram.cpp [hunk 6]
```diff
@@ -786,21 +786,34 @@ ACE_SOCK_Dgram::make_multicast_ifaddr6 (ipv6_mreq *ret_mreq,
...
+              iterations++;
```
Comment (mitza-oci) [inline ACE/ace/SOCK_Dgram.cpp:804]:
```suggestion
              ++iterations;
```

File: ACE/ace/SOCK_Dgram.cpp [hunk 7]
```diff
@@ -786,21 +786,34 @@ ACE_SOCK_Dgram::make_multicast_ifaddr6 (ipv6_mreq *ret_mreq,
...
+            }
+          else
+            {
+              break;
+            }
+        } while ((dwRetVal == ERROR_BUFFER_OVERFLOW) && (iterations < maxTries));
```
Comment (mitza-oci) [inline ACE/ace/SOCK_Dgram.cpp:810]:
```suggestion
        } while (dwRetVal == ERROR_BUFFER_OVERFLOW && iterations < maxTries);
```

File: ACE/ace/SOCK_Dgram.cpp [hunk 8]
```diff
@@ -819,14 +832,11 @@ ACE_SOCK_Dgram::make_multicast_ifaddr6 (ipv6_mreq *ret_mreq,
...
 
-#endif /* ACE_WIN32 */
+#else /* ACE_WIN32 */
 #ifndef ACE_LACKS_IF_NAMETOINDEX
-      if (lmreq.ipv6mr_interface == 0)
-        {
-          lmreq.ipv6mr_interface = ACE_OS::if_nametoindex (ACE_TEXT_ALWAYS_CHAR (net_if));
-        }
-
+      lmreq.ipv6mr_interface = ACE_OS::if_nametoindex(ACE_TEXT_ALWAYS_CHAR(net_if));
```
Comment (mitza-oci) [inline ACE/ace/SOCK_Dgram.cpp:837]:
```suggestion
      lmreq.ipv6mr_interface = ACE_OS::if_nametoindex (ACE_TEXT_ALWAYS_CHAR (net_if));
```

**Review by mitza-oci — CHANGES_REQUESTED** [review]

File: ACE/ace/SOCK_Dgram.cpp [hunk 9]
```diff
@@ -786,21 +786,34 @@ ACE_SOCK_Dgram::make_multicast_ifaddr6 (ipv6_mreq *ret_mreq,
...
-      // Initial call to determine actual memory size needed
-      ULONG bufLen = 0;
+      ULONG bufLen = 15000; // Initial size as per Microsoft
       char *buf = 0;
-      if (::GetAdaptersAddresses (AF_INET6, 0, 0, &tmp_addrs, &bufLen)
-          == ERROR_BUFFER_OVERFLOW)
+      ACE_NEW_RETURN (buf, char[bufLen], -1);
+      DWORD dwRetVal = 0;
+      ULONG iterations = 0;
+      const ULONG maxTries = 3;
```
Comment (mitza-oci) [inline ACE/ace/SOCK_Dgram.cpp:794]:
```suggestion
      ULONG const maxTries = 3;
```

**Review by mitza-oci — CHANGES_REQUESTED** [review]

**Review by mitza-oci — APPROVED** [review]

**Comment by mitza-oci** [comment]
@jwillemsen how do we see what actually failed?  The "Details" link doesn't show anything

**Comment by iguessthislldo** [comment]
> how do we see what actually failed? The "Details" link doesn't show anything

I have the same issue in #1357

**Comment by jwillemsen** [comment]
Strange, that worked in the past, maybe an issue on github actions currently, I restarted the failing checks

**Pull request merged**